# SPDX-License-Identifier: BSD-2-Clause
# OpenRadView: skip empty bands before bounds checks; retain MCT arrays until retargeted.
--- a/src/lib/openjp2/j2k.c
+++ b/src/lib/openjp2/j2k.c
@@ -5922,8 +5922,8 @@
             opj_mct_data_t *new_mct_records;
             l_tcp->m_nb_max_mct_records += OPJ_J2K_MCT_DEFAULT_NB_RECORDS;
 
-            new_mct_records = (opj_mct_data_t *) opj_realloc(l_tcp->m_mct_records,
-                              l_tcp->m_nb_max_mct_records * sizeof(opj_mct_data_t));
+            new_mct_records = (opj_mct_data_t *) opj_malloc(
+                                  l_tcp->m_nb_max_mct_records * sizeof(opj_mct_data_t));
             if (! new_mct_records) {
                 opj_free(l_tcp->m_mct_records);
                 l_tcp->m_mct_records = NULL;
@@ -5931,6 +5931,11 @@
                 l_tcp->m_nb_mct_records = 0;
                 opj_event_msg(p_manager, EVT_ERROR, "Not enough memory to read MCT marker\n");
                 return OPJ_FALSE;
+            }
+
+            if (l_tcp->m_nb_mct_records != 0) {
+                memcpy(new_mct_records, l_tcp->m_mct_records,
+                       l_tcp->m_nb_mct_records * sizeof(opj_mct_data_t));
             }
 
             /* Update m_mcc_records[].m_offset_array and m_decorrelation_array
@@ -5954,6 +5959,7 @@
                 }
             }
 
+            opj_free(l_tcp->m_mct_records);
             l_tcp->m_mct_records = new_mct_records;
             l_mct_data = l_tcp->m_mct_records + l_tcp->m_nb_mct_records;
             memset(l_mct_data, 0, (l_tcp->m_nb_max_mct_records - l_tcp->m_nb_mct_records) *
--- a/src/lib/openjp2/t2.c
+++ b/src/lib/openjp2/t2.c
@@ -1253,6 +1253,9 @@
     l_band = l_res->bands;
     for (bandno = 0; bandno < l_res->numbands; ++bandno, ++l_band) {
         opj_tcd_precinct_t *l_prc;
+        if (opj_tcd_is_band_empty(l_band)) {
+            continue;
+        }
         if (!(p_pi->precno < (l_band->precincts_data_size /
                               sizeof(opj_tcd_precinct_t)))) {
             opj_event_msg(p_manager, EVT_ERROR,
@@ -1263,10 +1266,6 @@
             return OPJ_FALSE;
         }
         l_prc = &(l_band->precincts[p_pi->precno]);
-
-        if (opj_tcd_is_band_empty(l_band)) {
-            continue;
-        }
 
         l_nb_code_blocks = l_prc->cw * l_prc->ch;
         l_cblk = l_prc->cblks.dec;
